File: server/src/controllers/languagemodels.py
from fastapi import APIRouter, HTTPException
from llm.llm_manager import LlmManager
from aif_types.languagemodels import ListLanguageModelsResponse, UpdateLmProviderRequest, ListLmProvidersResponse
from aif_types.llm import LlmFeature
from consts import ADMIN_CTRL_PREFIX
from utils.exception_utils import exceptionHandler


def create_admin_routers(llm_manager: LlmManager):
    router = APIRouter()

    @router.get(ADMIN_CTRL_PREFIX + "/languagemodels/", tags=["languagemodels"])
    async def list_languagemodels() -> ListLanguageModelsResponse:
        return exceptionHandler(lambda: llm_manager.list_languagemodels(LlmFeature.ALL))


    @router.get(ADMIN_CTRL_PREFIX + "/languagemodels/filter/{filter}", tags=["languagemodels"])
    async def list_languagemodels_with_filter(filter: LlmFeature) -> ListLanguageModelsResponse:
        if filter not in LlmFeature:
            raise HTTPException(status_code=400, detail="Invalid filter")
        return exceptionHandler(lambda: llm_manager.list_languagemodels(filter))


    @router.get(ADMIN_CTRL_PREFIX + "/languagemodels/providers", tags=["languagemodels"])
    async def listLmProviders() -> ListLmProvidersResponse:
        return exceptionHandler(lambda: llm_manager.listLmProviders())


    @router.post(ADMIN_CTRL_PREFIX + "/languagemodels/providers", tags=["languagemodels"])
    async def updateLmProvider(
        request: UpdateLmProviderRequest,
    ) -> ListLmProvidersResponse:
        return exceptionHandler(lambda: llm_manager.updateLmProvider(request))


    return router

File: packages/server-js/src/controllers/languagemodels.ts
import * as express from "express";
import { api, consts, StreamingUtils } from 'aifoundry-vscode-shared';
import ILmManager from "../lm/ILmManager";
import ResponseUtils from "../utils/ResponseUtils";
import RouterUtils from "../utils/RouterUtils";
import { HttpException } from "../exceptions";
import { ApiOutStream } from "../types/ApiOutStream";


export function registerAdminRoutes(router: express.Router, lmManager: ILmManager) {
    // Setup language model provider, for example, download Ollama to the host environment
    router.post(
        `${consts.ADMIN_CTRL_PREFIX}/languagemodels/setup`,
        RouterUtils.middlewares.jsonParser,
        (req, res) => {
            try {
                const request: api.SetupLmProviderRequest = req.body;
                lmManager.setupLmProvider(request, new ApiOutStream(res));    
            } catch (err) {
                ResponseUtils.handleException(res, err);
            }
        });

    // Get language model by feature filter, e.g. "embedding", "conversational"
    router.get(`${consts.ADMIN_CTRL_PREFIX}/languagemodels/filter/:filter`, (req, res) => {
        ResponseUtils.handler<api.ListLanguageModelsResponse>(res, async () => {
            const filter = req.params.filter as api.LlmFeature;
            if (api.LlmFeatures.indexOf(filter) === -1) {
                throw new HttpException(400, "Invalid filter");
            } else {
                return lmManager.listLanguageModels(filter);
            }
        });
    });

    // Download language model
    router.post(`${consts.ADMIN_CTRL_PREFIX}/languagemodels/crud/:lmproviderid/:id`, (req, res) => {
        try {
            lmManager.downloadLocalLanguageModel(req.params.lmproviderid, req.params.id, new ApiOutStream(res));
        } catch (err) {
            ResponseUtils.handleException(res, err);
        }
    });

    // Delete language model
    router.delete(`${consts.ADMIN_CTRL_PREFIX}/languagemodels/crud/:lmproviderid/:id`, (req, res) => {
        try {
            lmManager.deleteLocalLanguageModel(req.params.lmproviderid, req.params.id, new ApiOutStream(res));
        } catch (err) {
            ResponseUtils.handleException(res, err);
        }
    });

    // List language model provider
    router.get(`${consts.ADMIN_CTRL_PREFIX}/languagemodels/providers`, (req, res) => {
        const force = req.query[consts.QUERY_PARAM_FORCE] === "true";
        ResponseUtils.handler<api.ListLmProvidersResponse>(
            res,
            async () => lmManager.listLmProviders(force),
            ResponseUtils.maskListLmProvidersResponse,
        );
    });

    // Get language model provider by id
    router.get(`${consts.ADMIN_CTRL_PREFIX}/languagemodels/providers/:id`, (req, res) => {
        const force = req.query[consts.QUERY_PARAM_FORCE] === "true";
        ResponseUtils.handler<api.LmProviderInfoResponse>(
            res,
            async () => lmManager.getLmProvider(req.params.id, force),
            ResponseUtils.maskGetLmProviderInfoResponse,
        );
    });
    
    // Update language model provider
    router.post(
        `${consts.ADMIN_CTRL_PREFIX}/languagemodels/providers`,
        RouterUtils.middlewares.jsonParser,
        (req, res) => {
            ResponseUtils.handler<api.UpdateLmProviderResponse>(
                res,
                async () => lmManager.updateLmProviderInfo(req.body as api.UpdateLmProviderInfoRequest),
                ResponseUtils.maskUpdateLmProviderResponse,
            );
        }
    );

    // Update language model provider model configuration
    router.post(
        `${consts.ADMIN_CTRL_PREFIX}/languagemodels/providers/models`,
        RouterUtils.middlewares.jsonParser,
        (req, res) => {
            ResponseUtils.handler<api.UpdateLmProviderResponse>(
                res,
                async () => lmManager.updateLmProviderModel(req.body as api.UpdateLmProviderModelRequest),
                ResponseUtils.maskUpdateLmProviderResponse,
            );
        }
    );
}
